src/tokenizedSubaccounts/PPTSA.sol

```diff
@@ -0,0 +1,492 @@
+      maxLossOfOpenOptions + totalTradeMaxLoss
+        > baseBalance.multiplyDecimal(_getBasePrice()).multiplyDecimal($.ppParams.maxBuyPctOfTVL)
+    ) {
+      revert PPT_TradeTooLarge();
+    }
+  }
+
+  /////////////////
+  // Option Math //
+  /////////////////
+
+  function _validateTradeDetails(StrikeData memory lowerStrike, StrikeData memory higherStrike) internal view {
+    PPTSAStorage storage $ = _getPPTSAStorage();
+    int markCost = lowerStrike.markPrice.multiplyDecimal(lowerStrike.tradeAmount)
+      + higherStrike.markPrice.multiplyDecimal(higherStrike.tradeAmount);
+
+    int totalCostOfTrade = lowerStrike.tradePrice.toInt256().multiplyDecimal(lowerStrike.tradeAmount)
+      + higherStrike.tradePrice.toInt256().multiplyDecimal(higherStrike.tradeAmount);
+
+    if ($.ppParams.isLongSpread) {
+      if ($.ppParams.maxTotalCostTolerance < 1e18) {
+        revert PPT_InvalidCostTolerance();
+      } else if (totalCostOfTrade.abs() > markCost.abs().multiplyDecimal($.ppParams.maxTotalCostTolerance)) {
+        revert PPT_TotalCostOverTolerance();
+      }
+    } else {
+      if ($.ppParams.maxTotalCostTolerance > 1e18) {
+        revert PPT_InvalidCostTolerance();
+      } else if (totalCostOfTrade.abs() < markCost.abs().multiplyDecimal($.ppParams.maxTotalCostTolerance)) {
+        revert PPT_TotalCostBelowTolerance();
+      }
+    }
+
+    uint markValueToStrikeDiffRatio =
+      ((lowerStrike.markPrice - higherStrike.markPrice).abs().divideDecimal(higherStrike.strike - lowerStrike.strike));
+
+    if (
+      markValueToStrikeDiffRatio < $.ppParams.minMarkValueToStrikeDiffRatio
+        || markValueToStrikeDiffRatio > $.ppParams.maxMarkValueToStrikeDiffRatio
+    ) {
+      revert PPT_MarkValueNotWithinBounds();
+    }
+  }
+
+  function _getMarkPrice(IRfqModule.TradeData memory trade)
+    internal
+    view
+    returns (uint expiry, uint strike, uint markPrice)
+  {
+    PPTSAStorage storage $ = _getPPTSAStorage();
+    (uint optionExpiry, uint optionStrike, bool isCall) = OptionEncoding.fromSubId(trade.subId.toUint96());
+    if ($.ppParams.isCallSpread != isCall) {
+      revert PPT_WrongInputSpread();
+    }
+    if (block.timestamp >= optionExpiry) {
+      revert PPT_OptionExpired();
+    }
+    (uint callPrice, uint putPrice,) = _getOptionPrice(optionExpiry, optionStrike);
+    markPrice = isCall ? callPrice : putPrice;
+    return (optionExpiry, optionStrike, markPrice);
+  }
+
+  ///////////////////
+  // Account Value //
+  ///////////////////
+
+  function _getSubAccountStats() internal view returns (uint openSpreads, uint baseBalance, int cashBalance) {
+    BaseTSAAddresses memory tsaAddresses = getBaseTSAAddresses();
+    ISubAccounts.AssetBalance[] memory balances = tsaAddresses.subAccounts.getAccountBalances(subAccount());
+    for (uint i = 0; i < balances.length; i++) {
+      if (balances[i].asset == _getPPTSAStorage().baseCollateralAddresses.optionAsset) {
+        openSpreads = balances[i].balance.abs();
```

Comment:
🔴 This doesnt account for open positions on different strikes, and the code above assumes you're opening the same strikes too (which isn't enforced anywhere)

Reply:
Ah great catch. I wasn't sure if the subaccount that the TSA uses would use any other strikes other than the 2 we care about. Sounds like there might possibly be some more strikes. Well probably have to add some filtering logic

Reply:
In terms of opening the same strikes. I think that part is ok since we should have the same absolute value of two different strikes since we always require the two strikes to have the same amount * -1.

Reply:
Deciding for now that Ill just sum all of the > 0 balances. I think this will work with withdraws (no open spreads). and trades where we do one set of strikes one week and one set of strikes a different week (balances for settled spreads should be 0).